### PR TITLE
refactor: migrate Transaction, Enums, & AbstractDialect

### PR DESCRIPTION
**Pull request by ephys**
### Description Of Change

*Don't merge this in v6. Includes breaking changes*

Migrated `transaction.ts`, `table-hints.ts`, `query-types.ts`, `index-hints.ts`, & `dialects/abstract/index.ts`.

I also removed the hack in `operators.ts`

Note: I opted not to use any default exports in my refactor. This way, code that has not been migrated yet can be changed to
`const { Op } = require('./operators');` instead of the more verbose `const { default: Op } = require('./operators');`  
It's also easier to do `export * from` when we don't use default exports :)

Lots of changed files due to the export style change.

More important note: The typings for `ConnectionManager#getConnection` declared accepting `{ type: 'read' | 'write' }` as a parameter. It in fact accepted 'SELECT' instead of 'read'. I've changed it to 'read' to match typings. This is a breaking change for people using ConnectionManager directly.

### Todos

- [x] See which dialect fails

### CODE REVIEW

**Comment by WikiRik** [comment]
Could you document the breaking changes in the migration docs as well? https://github.com/sequelize/sequelize/blob/main/docs/manual/other-topics/upgrade-to-v7.md

Preferably some examples of the old behaviour and new behaviour so people see how to adopt it.

**Comment by ephys** [comment]
Done! :)

**Comment by ephys** [comment]
Before we merge this I'm still a bit confused when it comes to which branch needs to be targetted.

This should not land on v6, only on v7. Does that mean I should target the v7 branch or continue with `main`?

File: lib/dialects/abstract/index.ts
Comment (sdepold) [inline lib/dialects/abstract/index.ts:77]: Any reason why we don't inline this?

```
export class AbstractDialect {
  const supports: DialectSupports = { /* stuff here */ } 
}
```

File: lib/index-hints.ts
Comment (sdepold) [inline lib/index-hints.ts:8]: What are we shadowing?

**Comment by sdepold** [comment]
(I hope) We have agreed on:

- `main` to be the primary development branch for v7
- Any change to be introduced to v6 is either pointing to the master branch and cherry-picked to the v6 branch manually. Tracking happens by adding those changes to the v6 release project
- Any problem that only exists in the v6 code base because we have diverged too much at some point, can be fixed via a change pointing at v6 directly (which will immediately trigger a new release).

**Review by sdepold — DISMISSED** [review]
Amazing stuff!

File: lib/index-hints.ts
Comment (ephys) [reply lib/index-hints.ts:8]: I have no idea! I think it's no-shadow being broken with typescript. Should be fixed once we install `@typescript/no-shadow` instead

File: lib/dialects/abstract/index.ts
Comment (ephys) [reply lib/dialects/abstract/index.ts:77]: I copied this part over but my guess is the initial idea was to be able to access this object without instantiating the Dialect class.

Comment (sdepold) [reply lib/dialects/abstract/index.ts:77]: What makes more sense from your pov?

File: lib/index-hints.ts
Comment (sdepold) [reply lib/index-hints.ts:8]: Maybe just remove it?

File: lib/dialects/abstract/index.ts
Comment (ephys) [reply lib/dialects/abstract/index.ts:77]: I think adding a field to a prototype is prone to errors. I'd prefer assigning a frozen object to the constructor & the instance.

Let's find out if anything breaks if we do this.

File: lib/index-hints.ts
Comment (ephys) [reply lib/index-hints.ts:8]: I'd prefer to keep it for now. I'll try to see about writing a better eslint config & fix all issues in that PR :)

Speaking of which [I have a preset for eslint](https://github.com/ephys/eslint-config) that I've developed over the years initially based on airbnb. Could be a good starting point?

File: lib/dialects/abstract/index.ts
Comment (ephys) [reply lib/dialects/abstract/index.ts]: Done. `AbstractDialect.prototype.supports` still exists but it's now a getter that returns the static field

File: lib/index-hints.ts
Comment (sdepold) [reply lib/index-hints.ts:8]: I'm fine with any config really :) as long as my editor understands how to format the file for me.

Comment (sdepold) [reply lib/index-hints.ts:8]: I think we should format the entire code base as a one off thing once we settled on a config. Wanna file a pr?

Comment (ephys) [reply lib/index-hints.ts:8]: Sure :)

Comment (WikiRik) [reply lib/index-hints.ts:8]: I believe we wanted to wait until the TypeScript migration was done so we wouldn't have to keep any specific rules for vanilla JS and we could do a proper config at once. But we can try out the preset of @ephys if it doesn't give many issues with the current JS codebase

File: lib/dialects/abstract/index.ts
Comment (WikiRik) [inline lib/dialects/abstract/index.ts:12]: Do we need the comments on both parts? I would only keep these in the implementations

File: lib/dialects/db2/index.js
Comment (WikiRik) [inline lib/dialects/db2/index.js:12]: Can we only set the ones that are different from AbstractDialect here?

Comment (WikiRik) [reply lib/dialects/db2/index.js:12]: Also counts for the other dialects of course

File: lib/dialects/mysql/index.js
Comment (WikiRik) [inline lib/dialects/mysql/index.js:47]: I don't think this empty line is consistent with the other dialects

File: lib/index-hints.ts
Comment (ephys) [reply lib/index-hints.ts:8]: Well, only 12290 errors!

**Review by WikiRik — CHANGES_REQUESTED** [review]
Few small remarks, but overall looks good. Nice work!

File: lib/dialects/abstract/index.ts
Comment (WikiRik) [reply lib/dialects/abstract/index.ts]: I was about to say, we should refactor the usages of protoype out as much as possible in later PRs.

File: lib/index-hints.ts
Comment (WikiRik) [reply lib/index-hints.ts:8]: After running it with the --fix flag or before?

Comment (ephys) [reply lib/index-hints.ts:8]: There are some rules I'll disable for the JS codebase, but keep enabled for TS. This should ease the transition.

Comment (ephys) [reply lib/index-hints.ts:8]: Down to 900 errors. There are a lot of autofixed changes, which will be annoying for existing PRs. It's designed to minimise diff noise in the future so I think it's still worth it.

Comment (ephys) [reply lib/index-hints.ts:8]: 578 errors left but these are actual code smells that should be properly fixed (like using `==`, `new Array`, etc)

Comment (sdepold) [reply lib/index-hints.ts:8]: Hope you'll do that in another pr 😅

Comment (ephys) [reply lib/index-hints.ts:8]: Yeah :)

File: lib/dialects/mysql/index.js
Comment (ephys) [reply lib/dialects/mysql/index.js]: Removed!

File: lib/dialects/db2/index.js
Comment (ephys) [reply lib/dialects/db2/index.js:12]: Done. And updated the typings accordingly

File: lib/dialects/abstract/index.ts
Comment (ephys) [reply lib/dialects/abstract/index.ts:21]: I kept them in the typing only so they're shared by all implementations

File: lib/dialects/mssql/index.js
Comment (WikiRik) [inline lib/dialects/mssql/index.js:12]: I think this can be removed right?

Comment (ephys) [reply lib/dialects/mssql/index.js:12]: 🤦‍♀️

I added that to ensure both versions were deep equal, removing

**Review by WikiRik — DISMISSED** [review]

**Comment by ephys** [comment]
I'd like to see about merging https://github.com/sequelize/sequelize/pull/13930 first so I can lint these new files too

**Comment by ephys** [comment]
The new linter detected an issue in Transaction.ts: `cleanup` was not awaited in `commit` or `rollback`.

I don't know if this was voluntary but it could lead to an unhandled promise rejection error.

Also after commit hooks are awaited sequentially (https://github.com/sequelize/sequelize/pull/13922/files#diff-08c4998b82afd148218df2f2e7ce65983d0068469af7236efe0d15126f942269R83). I don't know if this is on purpose but I opted to not change it.

File: lib/index-hints.ts
Comment (ephys) [reply lib/index-hints.ts]: Shadowing resolved!

**Review by WikiRik — DISMISSED** [review]
One small suggestion, but the changes because of the updated linting still look good

**Review by WikiRik — APPROVED** [review]

**Pull request merged**
